### PR TITLE
sample: separate softmax and temperature transforms

### PR DESCRIPTION
**Pull request by ParthSareen**
Separating the transforms for readability

### CODE REVIEW

File: sample/transforms.go [hunk 1]
```diff
@@ -25,8 +25,17 @@ func (h *tokenHeap) Pop() any {
 	return x
 }
 
-// temperature applies scaling and softmax to the logits
+// temperature applies scaling to the logits
 func temperature(ts []token, temp float32) []token {
+	temp = max(temp, 1e-7)
```
Comment (jmorganca) [inline sample/transforms.go:30]: I don't mind the `max`. For readability we may want to do something like:

```
// avoid division by zero
if t < 1e-7 {
    t = 1e-7
}
```

File: sample/transforms_test.go [hunk 2]
```diff
@@ -32,9 +32,20 @@ func compareLogits(t *testing.T, name string, want []float32, got []token) {
 	}
 }
 
-func TestTemperatureAndSoftmax(t *testing.T) {
+func TestTemperature(t *testing.T) {
```
Comment (jmorganca) [inline sample/transforms_test.go:35]: Add tests for `0`

File: sample/transforms.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ParthSareen) [reply sample/transforms.go:30]: Can I comments instead?

File: sample/transforms.go [hunk 3]
```diff
@@ -25,8 +25,17 @@ func (h *tokenHeap) Pop() any {
...
+	for i := range ts {
+		ts[i].value = ts[i].value / temp
+	}
+	return ts
+}
+
+// softmax applies normalization to the logits
+func softmax(ts []token) []token {
```
Comment (jmorganca) [inline sample/transforms.go:38]: Let's add some tests for `softmax`

File: sample/transforms_test.go [hunk 4]
```diff
@@ -45,14 +56,11 @@ func TestTemperatureAndSoftmax(t *testing.T) {
...
-	// Check probabilities sum to 1
-	sum = 0.0
-	for _, token := range got {
-		sum += token.value
-	}
-	if math.Abs(float64(sum-1.0)) > 1e-6 {
-		t.Errorf("probabilities don't sum to 1: got %f", sum)
+	// Check relative ordering is preserved
+	for i := 1; i < len(got); i++ {
+		if (input[i] > input[i-1]) != (got[i].value > got[i-1].value) {
```
[the remark this answers is not in the document]
Comment (ParthSareen) [reply sample/transforms_test.go:61]: Actually we don't really need this

File: sample/samplers.go [hunk 5]
```diff
@@ -87,8 +87,8 @@ func (s *Sampler) sample(tokens []token) (token, error) {
 	// topK also sorts the tokens in descending order of logits
 	tokens = topK(tokens, s.topK)
 
-	// token logit values are updated to probabilities
 	tokens = temperature(tokens, s.temperature)
+	tokens = softmax(tokens)
```
Comment (jmorganca) [inline sample/samplers.go:91]: Any further thoughts on how we could avoid the `func transform(tokens []token) []token` interface that both mutates and returns a slice. This is more of a reminder, it can be in a follow up change. Perhaps we change the interface to

```func transform(tokens *[]token)``` and it can be called as

```softmax(&tokens)``` which visibly edits the slice

Comment (jmorganca) [reply sample/samplers.go:91]: And FWIW, returning a `[]token` is fine too, but doing so implies that `tokens` may not be mutated. If it ends up being the best scenario to keep it as-is we should add a comment to the ones that mutate arguments so callers know

Comment (ParthSareen) [reply sample/samplers.go:91]: Going to hack on this change in a follow up

File: sample/transforms_test.go (hunk 2 above)
Comment (ParthSareen) [reply sample/transforms_test.go:35]: Ah yes cool, let me add

File: sample/transforms.go (hunk 3 above)
Comment (ParthSareen) [reply sample/transforms.go:39]: Added some edge cases :D

File: sample/transforms_test.go [hunk 6]
```diff
@@ -32,27 +32,72 @@ func compareLogits(t *testing.T, name string, want []float32, got []token) {
...
-	got = temperature(toTokens(input), 1)
-	// Check probabilities sum to 1
-	sum = 0.0
-	for _, token := range got {
-		sum += token.value
+func TestSoftmax(t *testing.T) {
+	tests := []struct {
+		name  string
+		input []float32
+	}{
```
Comment (jmorganca) [inline sample/transforms_test.go:54]: Let's add a test with manually computed values

```
    name: "simple softmax",
    input: []float32{1, 2, 3},
    // optional expected field that tests the result
    expected: []float32{0.09003057, 0.24472848, 0.66524094},
```

**Review by jmorganca — APPROVED** [review]
